### PR TITLE
Basic CLI Implementation

### PR DESCRIPTION
**Pull request by MaxIsWell42**
Basic CLI completed with Boost program options. Includes help, version and build flags

### CODE REVIEW

**Comment by sethitow** [comment]
Please remove .ds_store from source control. This is a MacOS file that stores your desktop icon layout and isn't useful to other contributors.

File: example.cpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#include <boost/lambda/lambda.hpp>
```
Comment (sethitow) [inline example.cpp:1]: What is this file for?

File: src/main.cpp [hunk 2]
```diff
@@ -1,9 +1,50 @@
 #include <cstdlib>
 #include <iostream>
+#include <boost/program_options.hpp>
+#include <iterator>
```
Comment (sethitow) [inline src/main.cpp:4]: Please order standard library includes before 3rd party includes.
```suggestion
#include <iterator>

#include <boost/program_options.hpp>
```

File: src/main.cpp [hunk 3]
```diff
@@ -1,9 +1,50 @@
...
+
+using namespace std;
+namespace po = boost::program_options;
+
+
 
 int main(int argc, char *argv[])
 {
-    std::cout << "argc == " << argc << '\n';
+    // cout << "argc == " << argc << '\n';
```
Comment (sethitow) [inline src/main.cpp:16]: Please remove commented code

File: src/main.cpp [hunk 4]
```diff
@@ -1,9 +1,50 @@
...
+        // Get the input from the command line and store it in a variable map
+        po::variables_map vm;
+        po::store(parse_command_line(argc, argv, desc), vm);
+        po::notify(vm);
+
+        // Output based on command
+        if (vm.count("help"))
+            std::cout << desc << '\n';
+        else if (vm.count("version"))
+            // std::cout << "Current version: " << vm["age"].as<float>() << '\n';
```
Comment (sethitow) [inline src/main.cpp:35]: Please remove commented code.

File: src/main.cpp [hunk 5]
```diff
@@ -1,9 +1,50 @@
 #include <cstdlib>
 #include <iostream>
+#include <boost/program_options.hpp>
+#include <iterator>
+
+#include "parse.h"
+#include "compiling.h"
```
Comment (sethitow) [inline src/main.cpp:7]: To match the tense of `parse.h`, please rename this to `compile.h`.

File: src/compiling.h [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#include <iostream>
+using namespace std;
+
+int compiling() {
```
Comment (sethitow) [inline src/compiling.h:4]: Please move implementation to a .cpp file.

File: src/compiling.h [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#include <iostream>
+using namespace std;
```
Comment (sethitow) [inline src/compiling.h:2]: Please don't import the entire `std::` namespace. Prefix identifiers with `std::` to make it explicit where they're coming from, or import only the part of the namespace you need (if possible).

File: src/main.cpp [hunk 8]
```diff
@@ -1,9 +1,50 @@
...
+
+using namespace std;
```
Comment (sethitow) [inline src/main.cpp:9]: [same as above] Please don't import the entire std:: namespace. Prefix identifiers with std:: to make it explicit where they're coming from, or import only the part of the namespace you need (if possible).

File: src/parse.h [hunk 9]
```diff
@@ -0,0 +1,7 @@
+#include <iostream>
+using namespace std;
```
Comment (sethitow) [inline src/parse.h:2]: [same as above] Please don't import the entire std:: namespace. Prefix identifiers with std:: to make it explicit where they're coming from, or import only the part of the namespace you need (if possible).

File: src/main.cpp [hunk 10]
```diff
@@ -1,9 +1,50 @@
...
+    catch(exception& e) {
+        cerr << "error: " << e.what() << "\n";
+        return 1;
+    }
+    catch(...) {
+        cerr << "Exception of unknown type!\n";
+    }
 
-    return EXIT_SUCCESS;
+    return 0;
```
Comment (sethitow) [inline src/main.cpp:49]: It's generally bad practice to have "magic numbers" in code. Upon first read, it's not obvious what meaning "0" has. Please use a predefined constant such as `EXIT_SUCCESS` (from `cstdlib`) or, if you need a status that's not predefined, define it as a macro at the top of the program.

File: example.cpp (hunk 1 above)
Comment (sethitow) [inline example.cpp:1]: What is this file for?

File: src/parse.h [hunk 11]
```diff
@@ -0,0 +1,7 @@
...
+
+int parse() {
```
Comment (sethitow) [inline src/parse.h:4]: [same as above] Please move this implementation to a .cpp file.

File: src/main.cpp [hunk 12]
```diff
@@ -1,9 +1,50 @@
...
+        po::notify(vm);
+
+        // Output based on command
+        if (vm.count("help"))
+            std::cout << desc << '\n';
+        else if (vm.count("version"))
+            // std::cout << "Current version: " << vm["age"].as<float>() << '\n';
+            std::cout << "Current version: 0.1" << '\n';
+        else if (vm.count("build"))
+            std::cout << "And this is where I would put my build functions..." << '\n';
```
Comment (sethitow) [inline src/main.cpp:38]: You define a `compiling` function, but don't seem to call it anywhere. Did you intend to call it here?

File: src/main.cpp (hunk 10 above)
Comment (sethitow) [reply src/main.cpp:49]: Looking at this closer, you should probably return the result of whatever option the user chooses (i.e. compiling)

**Review by sethitow — CHANGES_REQUESTED** [review]
Hi Max, I left some feedback to address before merging.

**Comment by sethitow** [comment]
Can you please rename this PR to something related to the feature it adds?

File: example.cpp (hunk 1 above)
Comment (MaxIsWell42) [reply example.cpp:1]: It was for testing and debugging, forgot to remove

File: src/main.cpp (hunk 10 above)
Comment (MaxIsWell42) [reply src/main.cpp:49]: So it should just be a string describing what is happening?

File: src/main.cpp (hunk 12 above)
Comment (MaxIsWell42) [reply src/main.cpp:38]: Originally yes, but calling them here throws weird errors - expected unqualified id and expected expression. I'll bring them up more in-depth at the next meeting.

**Pull request merged**